compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -794,23 +805,192 @@ generateSrcPkgFromLf config pkg = do
         , "{-# OPTIONS_GHC -Wno-unused-imports -Wno-missing-methods #-}"
         ]
 
--- | Returns 'True' if an LF type contains a reference to an
--- old-style typeclass. See 'tconIsOldTypeclass' for more details.
-typeHasOldTypeclass :: Env -> LF.Type -> Bool
-typeHasOldTypeclass env = \case
-    LF.TVar _ -> False
-    LF.TCon tcon -> tconIsOldTypeclass env tcon
-    LF.TSynApp _ _ -> False
-        -- Type synonyms came after the switch to new-style
-        -- typeclasses, so we can assume there are no old-style
-        -- typeclasses being referenced.
-    LF.TApp a b -> typeHasOldTypeclass env a || typeHasOldTypeclass env b
-    LF.TBuiltin _ -> False
-    LF.TForall _ b -> typeHasOldTypeclass env b
-    LF.TStruct fields -> any (typeHasOldTypeclass env . snd) fields
-    LF.TNat _ -> False
-
--- | Determine whether a typecon refers to an old-style
+-- | A reference that can appear in a type or expression. We need to track
+-- all of these in order to control what gets hidden in data-dependencies.
+data Ref
+    = RTypeCon (LF.Qualified LF.TypeConName)
+        -- ^ necessary to track hidden types and old-style typeclasses
+    | RTypeSyn (LF.Qualified LF.TypeSynName)
+        -- ^ necessary to track hidden new-style typeclasses
+    | RValue (LF.Qualified LF.ExprValName)
+        -- ^ necessary to track hidden typeclass instances
+    deriving (Eq, Ord, Show, Generic)
+
+instance Hashable Ref
+
+refPackage :: Ref -> LF.PackageRef
+refPackage = \case
+    RTypeCon q -> LF.qualPackage q
+    RTypeSyn q -> LF.qualPackage q
+    RValue q -> LF.qualPackage q
+
+refModule :: Ref -> LF.ModuleName
+refModule = \case
+    RTypeCon q -> LF.qualModule q
+    RTypeSyn q -> LF.qualModule q
+    RValue q -> LF.qualModule q
+
+type RefGraph = HMS.HashMap Ref (Bool, [Ref])
+
+-- | Calculate the set of all references that should be hidden.
+buildHiddenRefMap :: Env -> HMS.HashMap Ref Bool
+buildHiddenRefMap env =
+    case leastFixedPointBy (||) refGraphList of
+        Left ref -> error
+            ("Internal error: missing reference in RefGraph "
+            <> show ref)
+        Right m -> m
+  where
+    refGraphList :: [(Ref, Bool, [Ref])]
+    refGraphList =
+        [ (ref, hidden, deps)
+        | (ref, (hidden, deps)) <- HMS.toList refGraph
+        ]
+
+    refGraph :: RefGraph
+    refGraph = foldl visitRef HMS.empty (rootRefs env)
+
+    visitRef :: RefGraph -> Ref -> RefGraph
+    visitRef !refGraph ref
+        | HMS.member ref refGraph
+            = refGraph -- already in the map
+        | ref == RTypeCon erasedTCon
+            = HMS.insert ref (True, []) refGraph -- Erased is always erased
+        | refModule ref == LF.ModuleName ["DA", "Generics"]
```

Comment:
Can we add a comment why `Erased` is not sufficient to throw out the problematic stuff from `DA.Generics`?

Reply:
I'll add a comment. It's because `GenConvertible` causes kind errors in the back and forth translation, without actual  references to `Erased`.

Reply:
FWIW, `GenConvertible` is killed in https://github.com/digital-asset/daml/pull/4839. Although I don’t really understand why `GenConvertible` causes issues here since it’s not used in the project you tested on afaik and we have other tests that use `daml-stdlib`. Either way, happy to keep this code if we add a comment.